src/Compilers/Core/Portable/MetadataReference/AssemblyIdentityMap.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Implements a map from an assembly identity to a value. The map allows to look up the value by an identity
+    /// that either exactly matches the original identity key, or corresponds to a key with the lowest version among identities
+    /// with higher version than the requested identity key.
+    /// </summary>
+    internal sealed class AssemblyIdentityMap<TValue>
+    {
+        private readonly Dictionary<string, OneOrMany<KeyValuePair<AssemblyIdentity, TValue>>> _map;
+
+        public AssemblyIdentityMap()
+        {
+            _map = new Dictionary<string, OneOrMany<KeyValuePair<AssemblyIdentity, TValue>>>(AssemblyIdentityComparer.SimpleNameComparer);
+        }
+
+        public bool Contains(AssemblyIdentity identity, bool allowHigherVersion = true)
+        {
+            TValue value;
+            return TryGetValue(identity, out value, allowHigherVersion);
+        }
+
+        public bool TryGetValue(AssemblyIdentity identity, out TValue value, bool allowHigherVersion = true)
+        {
+            OneOrMany<KeyValuePair<AssemblyIdentity, TValue>> sameName;
+            if (_map.TryGetValue(identity.Name, out sameName))
+            {
+                int minHigherVersionCandidate = -1;
+
+                for (int i = 0; i < sameName.Count; i++)
+                {
+                    AssemblyIdentity currentIdentity = sameName[i].Key;
+
+                    if (AssemblyIdentity.EqualIgnoringNameAndVersion(currentIdentity, identity))
```

Comment:
When the assemblies have the same name and version but different public key or culture.